### PR TITLE
utils/pkg/win.py Introducing new library to gather windows software/component info

### PR DESCRIPTION
**Pull request by damon-atkins**
### What does this PR do?
Adds utils/pkg/win.py to the 2017.7 (now develop) release. This collects an extensive amount of information about software and components installed on a windows system.
See sample data below

In 2017.7 this is independent code. i.e their is no part of salt which currently uses this code and no plans for any part of 2017.7 to use this code.  This code has been targeted for 2017.7 so that it can be made available for testing by a wider community of users. Users testing this code will run the library directly outside of salt.
```
python -B C:\salt\bin\Lib\site-packages\salt\utils\pkg\win.py
usage: C:\salt\bin\Lib\site-packages\salt\utils\pkg\win.py <detail|list> <system|system+user>
```

The aim is for the library to be used in a future release of salt e.g. 2018.X However in order for this to happen, there needs to be tested as much as possible to find any bugs.

Following is sample data

```
   "adobe systems inc\\\\adobe shockwave player 12.2\\\\soft32": {
        "arch": "x86",
        "vendor": "Adobe Systems, Inc",
        "version": {
            "12.2.9.199": {
                "install_date": "2017-07-02T00:00:00",
                "install_date_time_t": 1498917600,
                "size": 57563,
                "url": "http://www.adobe.com",
                "win_advertise_flags": 388,
                "win_display_name": "Adobe Shockwave Player 12.2",
                "win_install_location": "c:\\Windows\\SysWOW64\\Adobe\\",
                "win_install_src_dir": "c:\\salt\\var\\cache\\salt\\minion\\extrn_files\\base\\fpdownload.macromedia.com\\get\\shockwave\\default\\english\\win95nt\\latest\\",
                "win_installer_type": "winmsi",
                "win_instance_type": 0,
                "win_language": "en_US",
                "win_package_code": "{91110B3D-99F1-4271-9382-ED625D4CD606}",
                "win_package_name": "sw_lic_full_installer.msi",
                "win_product_code": "{39EA6AA6-F891-4D70-867D-839DA49948D2}",
                "win_redeployment_flags": 3,
                "win_uninstall_cmd": "MsiExec.exe /X{39EA6AA6-F891-4D70-867D-839DA49948D2}",
                "win_upgrade_code": "{2DEDA10C-F284-4C17-9421-E2391D6E8372}",
                "win_version_binary": "12.2.9",
                "win_version_display": "12.2.9.199",
                "win_version_user": "12.2.9.199"
            }
        },
        "win_system_component": false,
        "win_version_src": "display-version"
    }, 
    "microsoft corporation\\\\microsoft office shared setup metadata mui (english) 2010\\\\comp": {
        "arch": "x64", 
        "vendor": "Microsoft Corporation", 
        "version": {
            "14.0.7015.1000": {
                "install_date": "2017-03-09T00:00:00", 
                "install_date_time_t": 1488978000, 
                "size": 1724, 
                "win_advertise_flags": 388, 
                "win_display_name": "Microsoft Office Shared Setup Metadata MUI (English) 2010", 
                "win_install_location": "C:\\Program Files\\Microsoft Office\\", 
                "win_install_src_dir": "C:\\MSOCache\\All Users\\{90140000-0115-0409-1000-0000000FF1CE}-C\\", 
                "win_installer_type": "winmsi", 
                "win_instance_type": 0, 
                "win_language": "en_US", 
                "win_package_code": "{D310FF6B-7B42-4047-BAE4-767EBC438D51}", 
                "win_package_name": "OfficeMUISet.msi", 
                "win_patches": [
                    "Update for Microsoft Office 2010 (KBNNNNNNN) 64-Bit Edition", 
                    "Service Pack 2 for Microsoft Office 2010 (KBNNNNNNN) 64-Bit Edition"
                ], 
                "win_product_code": "{90140000-0115-0409-1000-0000000FF1CE}", 
                "win_redeployment_flags": 3, 
                "win_uninstall_cmd": "MsiExec.exe /X{90140000-0115-0409-1000-0000000FF1CE}", 
                "win_upgrade_code": "{00140000-0115-0000-1000-0000000FF1CE}", 
                "win_version_binary": "14.0.7015", 
                "win_version_display": "14.0.7015.1000", 
                "win_version_user": "14.0.7015.1000"
            }
        }, 
        "win_system_component": true, 
        "win_version_src": "display-version"
    }, 
```

### Previous Behavior
None

### New Behavior
None, as its not called by salt and not loaded by any modules.

### Commits signed with GPG?
No

### CODE REVIEW

**Comment by damon-atkins** [comment]
Lint errors `Incompatible Python 3 code found.`  are wrong, this code works under python 3

**Comment by cachedout** [comment]
Hi @damon-atkins 

Is your thinking here that we would only have users test this? Can automated tests be added for this?

**Comment by damon-atkins** [comment]
Hi @cachedout, its just targeted at the 2017 release so people can run it by hand, and report back issues, before it gets used in windows package management as the source of installed software/components.  If its easy for people to test with, should be able to get a few people to test it.

Would like to do tests latter, but need some help pointing me at some existing tests. Not sure how to go amount emulating the windows registry.    
Cheers

**Comment by damon-atkins** [comment]
Can someone look at the incorrect Python 3 lint errors.  ? @rallytime ?

**Comment by rallytime** [comment]
@damon-atkins The python 3 errors from our linter are harder to find than the normal ones. They're detailed in the console output here: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/16300/console

**Comment by damon-atkins** [comment]
Hi @rallytime  why do I need to put list in front of zip  The code works fine like it is. i.e. it returns a tuple of four items each item is a list.     Why do I need to turn my tuple into a "4 item list" with each item containing a list
Cheers

**Comment by damon-atkins** [comment]
Would appreciate if this is in the next release 2017.3.   Lint issues are false.

**Comment by damon-atkins** [comment]
re-run lint

**Comment by damon-atkins** [comment]
Go Go Jenkins!

**Comment by damon-atkins** [comment]
Go Go Jenkins!

**Comment by damon-atkins** [comment]
Hi @rallytime why do I need to put list in front of zip The code works fine like it is. i.e. it returns a tuple of four items each item is a list. Why do I need to turn my tuple into a "4 item list" with each item containing a list.
Hope the saltconf went well.
Cheers
```
14:16:46 -                    reg_key_all, _, _, _ = zip(*win32api.RegEnumKeyEx(handle))  # pylint: disable=no-member
14:16:46 +                    reg_key_all, _, _, _ = list(zip(*win32api.RegEnumKeyEx(handle)))  # pylint: disable=no-member
```

**Comment by rallytime** [comment]
Hi @damon-atkins - I am not entirely sure why that lint check is there. I believe discussion of that has come up in the past. Perhaps @s0undt3ch has some insight on that particular check.

In the mean time, we won't be able to merge this addition until lint is happy. :) Since this works as needed, you can add a pylint disable.

Also, can you move this to the `develop` branch please? You are adding a new feature and `2017.7` is bug-fix only.

**Comment by damon-atkins** [comment]
Hi @cachedout  this is ready to be merged.

**Comment by damon-atkins** [comment]
Hi @rallytime the only way to run this new code is talked about in the introduction.  It does not hook into any existing salt code. Any hooks will be added into development. Its target at 2017 so that its available for people to run by hand from the next 2017 release onward, get some text output and then report issues which need fixing, before its used in a release such as 2018.

**Comment by rallytime** [comment]
Hi @damon-atkins My preference is still to gate this in `develop` since this is a new feature. I know you'd like to get this in sooner, but unfortunately, we have to draw a line in the sand somewhere. Since it is possible to distribute utils files dynamically, it would still be relatively easy to start testing this sooner as you suggest. See [these docs](https://docs.saltstack.com/en/latest/ref/file_server/dynamic-modules.html) for more information.

**Comment by damon-atkins** [comment]
OK I will change the branch, I have a significant amount of code developed which was developed at the same time to use this library. But it needs lot of testing before I am willing to release the rest of the code.

**Comment by damon-atkins** [comment]
Go Go Jenkins!

**Comment by rallytime** [comment]
Thank you @damon-atkins - Let us know when this is ready to review and we can get the folks involved with Windows to help review.

**Comment by damon-atkins** [comment]
Hi @rallytime it can be review now.

**Comment by morganwillcock** [comment]
I've given it a go (this all seems to be useful stuff) and noted these issues:

- Missed identification of MSI installer type. It always searches for 'msiexec.exe' in the uninstall string, but in some cases the file extension is not present, so it also needs to match on 'msiexec':
<pre>"adobe systems incorporated.\\\\adobe community help\\\\soft32": {
    "arch": "x86", 
    "vendor": "Adobe Systems Incorporated.", 
    "version": {
        "3.4.980": {
            "install_date": "2014-08-19T11:20:12", 
            "install_date_time_t": 1408443612, 
            "win_display_name": "Adobe Community Help", 
            "win_install_location": "C:\\Program Files (x86)\\Adobe\\Adobe Help\\", 
            "win_installer_type": "unknown", 
            "win_product_code": "chc.4875E02D9FB21EE389F73B8D1702B320485DF8CE.1", 
            "win_uninstall_cmd": "msiexec /qb /x {3521BDBD-D453-5D9F-AA55-44B75D214629}", 
            "win_version_display": "3.4.980", 
            "win_version_user": "3.4.980"
        }
    }, 
    "win_system_component": false, 
    "win_version_src": "display-version"
}</pre>
- Pkg IDs / key names containing 'NoValue'. The pkg ID being built from data that might not be set means at best you might have to start using 'NoValue' in key names, and at worst you will get name collisions. Perhaps as there is already a function to return the registry path to the uninstall entry, it would be best to use that? On a lesser note, I wasn't sure what the difference was between 'soft' and 'comp' (I initially assumed it was short for 'computer') before I realised you were separating items which are flagged as system components. Since you have this available as 'win_system_component' I would say that it only confuses the key to have it present, especially since an upgrade may reuse other values or paths, but flip the system component flag, which would then change the key name.
<pre>"NoValue\\\\bde 5.11 for sims workstation setup\\\\soft32": {
    "arch": "x86", 
    "version": {
        "0.0.0.0.0": {
            "install_date": "2014-08-14T14:06:33", 
            "install_date_time_t": 1408021593, 
            "win_display_name": "BDE 5.11 for SIMS Workstation Setup", 
            "win_installer_type": "installshield", 
            "win_product_code": "{0FFC8EC3-38F4-4A3B-9D2D-6A56DDF4C97C}", 
            "win_uninstall_cmd": "RunDll32 C:\\PROGRA~2\\COMMON~1\\INSTAL~1\\engine\\6\\INTEL3~1\\Ctor.dll,LaunchSetup \"C:\\Program Files (x86)\\InstallShield Installation Information\\{0FFC8EC3-38F4-4A3B-9D2D-6A56DDF4C97C}\\Setup.exe\" -l0x9 ", 
            "win_version_user": "0.0.0.0.0"
        }
    }, 
    "win_system_component": false, 
    "win_version_src": "none"
}

"microsoft\u0000p\\\\NoValue\\\\soft32": {
    "arch": "x86", 
    "vendor": "Microsoft\u0000P", 
    "version": {
        "0.0.0.0.0": {
            "install_date": "2017-10-26T10:10:21", 
            "install_date_time_t": 1509009021, 
            "win_display_name": "", 
            "win_installer_type": "unknown", 
            "win_parent_name": "Microsoft Office Professional Plus 2013", 
            "win_parent_pkg_uid": "Office15.PROPLUS", 
            "win_product_code": "{90150000-0011-0000-0000-0000000FF1CE}_Office15.PROPLUS_{439F5167-ECC8-4ACB-9590-C0E7141EAAD3}", 
            "win_uninstall_cmd": "\"C:\\Program Files (x86)\\Common Files\\Microsoft Shared\\OFFICE15\\Oarpmany.exe\" /removereleaseinpatch \"{90150000-0011-0000-0000-0000000FF1CE}\" \"{439F5167-ECC8-4ACB-9590-C0E7141EAAD3}\" \"1033\" \"0\"", 
            "win_uninstall_no_remove": true, 
            "win_version_user": "0.0.0.0.0"
        }
    }, 
    "win_system_component": false, 
    "win_version_src": "none"
}</pre>
Traceback when searching for software in the context of a user. I think it was trying to open 'S-1-5-19\Software\Microsoft\Windows\CurrentVersion\Uninstall' which maps to 'NT Authority'.
<pre>[DEBUG]: Processing user software.. please wait
Traceback (most recent call last):
  File "C:\salt\bin\Lib\site-packages\salt\utils\pkg\win.py", line 1286, in <module>
    __main()
  File "C:\salt\bin\Lib\site-packages\salt\utils\pkg\win.py", line 1282, in __main
    print('Time Taken: {}'.format(timeit.timeit(run, number=1)))  # pylint: disable=superfluous-parens
  File "C:\salt\bin\lib\timeit.py", line 237, in timeit
    return Timer(stmt, setup, timer).timeit(number)
  File "C:\salt\bin\lib\timeit.py", line 202, in timeit
    timing = self.inner(it, self.timer)
  File "C:\salt\bin\lib\timeit.py", line 100, in inner
    _func()
  File "C:\salt\bin\Lib\site-packages\salt\utils\pkg\win.py", line 1278, in run
    pkg_list = WinSoftware(user_pkgs=user_pkgs, version_only=version_only)
  File "C:\salt\bin\Lib\site-packages\salt\utils\pkg\win.py", line 684, in __init__
    self.__get_software_details(user_pkgs=user_pkgs)
  File "C:\salt\bin\Lib\site-packages\salt\utils\pkg\win.py", line 1242, in __get_software_details
    win32con.KEY_READ)
pywintypes.error: (2, 'RegOpenKeyEx', 'The system cannot find the file specified.')</pre>

**Comment by damon-atkins** [comment]
I will have a look at it.   NoValue means field was found to be blank and it contains capital N and V.
The default pkg id format is `Publisher\\DisplayName\\<Type>` forced to lower case except for NoValue.
The only field which will see NoValue is Publisher. Type is one of soft|comp|soft32|comp32     Given how lose standard is the aim is to find a unique id. Component and Software can have the same name.   https://en.wikipedia.org/wiki/Windows_Installer#Components   97% of the time people will only deal soft software not software components.

Thanks for the feedback.
I will look into the other issues.

**Comment by morganwillcock** [comment]
As far as I know (but might be wrong), the components referenced here:
[https://en.wikipedia.org/wiki/Windows_Installer#Components](https://en.wikipedia.org/wiki/Windows_Installer#Components)
...are talking about components in the context of an MSI package, where it is the terminology for a reference to an item managed by that package. The flag for 'system component'  can be set for any installer type and just means that the uninstall entry is not shown in the Control Panel / Settings app.

**Comment by damon-atkins** [comment]
Bugs found so far should be fixed, this needs testing on a wide range of computers to ensure it is bug free. The different the better.  

Ping @TheBigBear @cheribral @DrMerlin @gudipudipradeep @stjack99 @Mark3K @jf @andrebron @rmarcinik @jefftucker @highlyunavailable

**Comment by damon-atkins** [comment]
Hi @morganwillcock can you test again and approve pls, so it gets into Oxygen.

**Comment by morganwillcock** [comment]
I don't have write access, so my review isn't going to push it through, but I still have concerns about the key value (unless the key is just used for demo purposes when printing all data):

<pre>"NoValue\\\\bde 5.11 for sims workstation setup\\\\soft32": {
	"arch": "x86", 
	"version": {
		"0.0.0.0.0": {
			"install_date": "2014-08-14T14:06:33", 
			"install_date_time_t": 1408021593, 
			"win_display_name": "BDE 5.11 for SIMS Workstation Setup", 
			"win_installer_type": "installshield", 
			"win_product_code": "{0FFC8EC3-38F4-4A3B-9D2D-6A56DDF4C97C}", 
			"win_uninstall_cmd": "RunDll32 C:\\PROGRA~2\\COMMON~1\\INSTAL~1\\engine\\6\\INTEL3~1\\Ctor.dll,LaunchSetup \"C:\\Program Files (x86)\\InstallShield Installation Information\\{0FFC8EC3-38F4-4A3B-9D2D-6A56DDF4C97C}\\Setup.exe\" -l0x9 ", 
			"win_version_user": "0.0.0.0.0"
		}
	}, 
	"win_system_component": false, 
	"win_version_src": "none"
}</pre>

Plus I think you are misidentifying anything with the system component flag set as being a Windows component. I think it just has an important sounding name but can be used for anything, so I don't think that should be part of the namespace of your key. As an example, this is [MIT Scratch](https://scratch.mit.edu/) which is just a programming tool made in Adobe Air, so not providing anything at a system level:

<pre>"mit media lab\\\\scratch 2 offline editor\\\\comp32": {
	"arch": "x86", 
	"vendor": "MIT Media Lab", 
	"version": {
		"255": {
			"install_date": "2016-01-05T00:00:00", 
			"install_date_time_t": 1451952000, 
			"size": 75259, 
			"win_advertise_flags": 388, 
			"win_display_name": "Scratch 2 Offline Editor", 
			"win_install_location": "C:\\Program Files (x86)\\Scratch 2", 
			"win_install_src_dir": "C:\\Users\\Admin\\AppData\\Local\\Temp\\AIR8BE9.tmp\\Scratch 2\\", 
			"win_installer_type": "winmsi", 
			"win_instance_type": 0, 
			"win_language": "en_US", 
			"win_package_code": "{DCEB8C84-8C31-0B01-CA35-3FD484C953E9}", 
			"win_package_name": "setup.msi", 
			"win_product_code": "{DCEB8C84-8C31-0B01-CA35-3FD484C953E9}", 
			"win_redeployment_flags": 3, 
			"win_uninstall_cmd": "MsiExec.exe /I{DCEB8C84-8C31-0B01-CA35-3FD484C953E9}", 
			"win_upgrade_code": "{F77CC691-768E-90D2-2DC3-89BED9EB6052}", 
			"win_version_binary": "255.0.0", 
			"win_version_display": "255", 
			"win_version_user": "255"
		}
	}, 
	"win_system_component": true, 
	"win_version_src": "display-version"
}</pre>

Also, got another traceback when checking users:

<pre>[DEBUG]: Processing user software.. please wait
[DEBUG]: Not Found S-1-5-19\Software\Microsoft\Windows\CurrentVersion\Uninstall
[DEBUG]: Not Found S-1-5-20\Software\Microsoft\Windows\CurrentVersion\Uninstall
[DEBUG]: Software/Component Not Found in Products section of registry key_guid: '{1197d2bb-6cf8-488a-b994-d5bf6d7efe7b}', sid: 'S-1-5-21-1192708525-2215310883-282776303-500', use_32bit: 'False'
Traceback (most recent call last):
  File "C:\Salt\bin\Lib\site-packages\salt\utils\pkg\win.py", line 1295, in <module>
    __main()
  File "C:\Salt\bin\Lib\site-packages\salt\utils\pkg\win.py", line 1291, in __main
    print('Time Taken: {}'.format(timeit.timeit(run, number=1)))  # pylint: disable=superfluous-parens
  File "C:\salt\bin\lib\timeit.py", line 237, in timeit
    return Timer(stmt, setup, timer).timeit(number)
  File "C:\salt\bin\lib\timeit.py", line 202, in timeit
    timing = self.inner(it, self.timer)
  File "C:\salt\bin\lib\timeit.py", line 100, in inner
    _func()
  File "C:\Salt\bin\Lib\site-packages\salt\utils\pkg\win.py", line 1287, in run
    pkg_list = WinSoftware(user_pkgs=user_pkgs, version_only=version_only)
  File "C:\Salt\bin\Lib\site-packages\salt\utils\pkg\win.py", line 687, in __init__
    self.__get_software_details(user_pkgs=user_pkgs)
  File "C:\Salt\bin\Lib\site-packages\salt\utils\pkg\win.py", line 1260, in __get_software_details
    self.__collect_software_info(sid, reg_key, False)
  File "C:\Salt\bin\Lib\site-packages\salt\utils\pkg\win.py", line 1014, in __collect_software_info
    username = self.__sid_to_username(sid)
  File "C:\Salt\bin\Lib\site-packages\salt\utils\pkg\win.py", line 874, in __sid_to_username
    win32api.NameUserPrincipal)  # pylint: disable=no-member
pywintypes.error: (1332, 'TranslateName', 'No mapping between account names and security IDs was done.')</pre>

File: salt/utils/pkg/win.py [hunk 1]
```diff
@@ -0,0 +1,1295 @@
...
+:maintainer: Salt Stack <https://github.com/saltstack>
+:codeauthor: Damon Atkins <https://github.com/damon-atkins>
+:maturity: new
+:depends: pywin32, six
+:platform: windows
+
+Known But install_date may not match Control Panel\Programs\Programs and Features
+
+'''
+
```
Comment (twangboy) [inline salt/utils/pkg/win.py]: Too many blank lines?

File: salt/utils/pkg/win.py [hunk 2]
```diff
@@ -0,0 +1,1295 @@
...
+        Returns:
+            dict or list: List if ``version_only`` is ``True`` otherwise dict
+
+        """
+        if pkg_id in self.__reg_software:
+            return self.__reg_software[pkg_id]
+        else:
+            raise KeyError(pkg_id)
+
+    def __iter__(self):
```
Comment (twangboy) [inline salt/utils/pkg/win.py]: Docstring?

File: salt/utils/pkg/win.py [hunk 3]
```diff
@@ -0,0 +1,1295 @@
...
+
+    @staticmethod
+    def __oldest_to_latest_version(ver1, ver2):
+        '''Used for sorting version numbers oldest to latest
+        '''
+        return 1 if LooseVersion(ver1) > LooseVersion(ver2) else -1
+
+    @staticmethod
+    def __latest_to_oldest_version(ver1, ver2):
+        ''''Used for sorting version numbers, latest to oldest
```
Comment (twangboy) [inline salt/utils/pkg/win.py]: Extra `'`

File: salt/utils/pkg/win.py [hunk 4]
```diff
@@ -0,0 +1,1295 @@
...
+        if isinstance(pkg_data, list):
+            # raw data is 'pkgid': [sorted version list]
+            return pkg_data  # already sorted oldest to newest
+
+        # Must be a dict or OrderDict, and contain full details
+        installed_versions = list(pkg_data.get('version').keys())
+        return sorted(installed_versions, key=cmp_to_key(self.__oldest_to_latest_version))
+
+    def pkg_version_latest(self, pkg_id):
+        """Returns a package latest version installed out of all the version
```
Comment (twangboy) [inline salt/utils/pkg/win.py]: versions

File: salt/utils/pkg/win.py [hunk 5]
```diff
@@ -0,0 +1,1295 @@
...
+            pkg_id (str): Package Id of the software/component.
+
+        Returns:
+            str: Latest/Newest version number installed.
+
+        """
+        return self.pkg_version_list(pkg_id)[-1]
+
+    def pkg_version_oldest(self, pkg_id):
+        """Returns a package oldest version installed out of all the version
```
Comment (twangboy) [inline salt/utils/pkg/win.py]: versions

File: salt/utils/pkg/win.py [hunk 6]
```diff
@@ -0,0 +1,1295 @@
...
+
+        Returns:
+            str: Oldest version number installed.
+
+        """
+        return self.pkg_version_list(pkg_id)[0]
+
+    @staticmethod
+    def __sid_to_username(uid):
+        """Provided with a valid Windows User SID and returns a Username
```
Comment (twangboy) [inline salt/utils/pkg/win.py]: Provided with a valid Windows User Sid (uid), return a Username


Consider using/improving the `get_name` function from `salt.utils.win_dacl`
https://github.com/saltstack/salt/blob/develop/salt/utils/win_dacl.py#L1097

File: salt/utils/pkg/win.py [hunk 7]
```diff
@@ -0,0 +1,1295 @@
...
+            # invalid. e.g. S-1-5-19 which is a local account
+            if exc.winerror == winerror.ERROR_NO_SUCH_DOMAIN or exc.winerror == winerror.ERROR_INVALID_DOMAINNAME:
+                return '{0}@{1}'.format(name.lower(), domain.lower())
+            else:
+                raise
+        return user_principal
+
+    def __software_to_pkg_id(self, publisher, name, iscomponent, is32bit):
+        """Determine the Package ID of a software/component using the
+        software/component ``publisher``, ``name``, weather its a software or a
```
Comment (twangboy) [inline salt/utils/pkg/win.py]: whether it's software or a component

File: salt/utils/pkg/win.py [hunk 8]
```diff
@@ -0,0 +1,1295 @@
...
+        component, and if its 32bit or 64bit archiecture.
+
+        Args:
+            publisher (str): Publisher of the software/component.
+            name (str): Name of the software.
+            iscomponent (bool): True if package is a component.
+            is32bit (bool): True if the software/component is 32bit architecture.
```
Comment (twangboy) [inline salt/utils/pkg/win.py]: `is_component`
`is_32bit`

File: salt/utils/pkg/win.py [hunk 9]
```diff
@@ -0,0 +1,1295 @@
...
+            pkg_id = self.__pkg_obj.to_pkg_id(publisher, name, iscomponent, is32bit)
+            if pkg_id:
+                return pkg_id
+
+        return default_pkg_id
+
+    def __version_capture_slp(self, pkg_id, version_binary, version_display, display_name):
+        """This returns version and where the version string came from. based on
+        instructions under version_capture, if version_capture missing defaults to
+        display-version
```
Comment (twangboy) [inline salt/utils/pkg/win.py]:
```
'''
This returns the version and where the version string came from based on
instructions under ``version_capture``. If ``version_capture`` is missing
it defaults to display-version
```

File: salt/utils/pkg/win.py [hunk 10]
```diff
@@ -0,0 +1,1295 @@
...
+            elif src != 'use-default':
+                raise ValueError(
+                    'version capture within object \'{0}\' failed '
+                    'for pkg id: \'{1}\' it returned \'{2}\' \'{3}\' '
+                    '\'{4}\''.format(str(self.__pkg_obj), pkg_id, version_str, src, version_user_str)
+                    )
+
+        # If self.__pkg_obj.version_capture() not defined defaults to using
+        # version_display and if not valid then use version_binary, and as a last
+        # result provided the version 0.0.0.0.0 to indicate version string was not determined.
```
Comment (twangboy) [inline salt/utils/pkg/win.py]: provide

File: salt/utils/pkg/win.py [hunk 11]
```diff
@@ -0,0 +1,1295 @@
...
+        # if NoRemove exists we would expect their to be no UninstallString
+        uninstall_no_remove = reg_soft_info.is_install_true('NoRemove')
+        uninstall_string = reg_soft_info.get_install_value('UninstallString')
+        uninstall_quiet_string = reg_soft_info.get_install_value('QuietUninstallString')
+        windows_installer = reg_soft_info.is_install_true('WindowsInstaller')
+        system_component = reg_soft_info.is_install_true('SystemComponent')
+        publisher = reg_soft_info.get_install_value('Publisher', wanted_type='str')
+
+        # UninstallString is optional if the installer is "windows installer"/MSI
+        # However for add to appear in Control-Panel -> Program and Features -> Uninstall or change a program
```
Comment (twangboy) [inline salt/utils/pkg/win.py]: for it to appear

File: salt/utils/pkg/win.py [hunk 12]
```diff
@@ -0,0 +1,1295 @@
...
+        # the UninstallString needs to be set or ModifyString set
+        # if (uninstall_string is None and uninstall_quiet_string is None) and (not windows_installer):
+        #    return
+
+        # Question: If uninstall string is not set and windows_installer should we set it
+        # Question: if uninstall_quite is not set .......
```
Comment (twangboy) [inline salt/utils/pkg/win.py]: quiet

File: salt/utils/pkg/win.py [hunk 13]
```diff
@@ -0,0 +1,1295 @@
...
+        # log.trace('%s\\%s ver:%s src:%s', username or 'SYSTEM', pkg_id, version_text, version_src)
+
+        if username:
+            dict_key = '{};{}'.format(username, pkg_id)  # Use ; as its not a valid hostnmae char
+        else:
+            dict_key = pkg_id
+
+        # Guessing the architecture http://helpnet.flexerasoftware.com/isxhelp21/helplibrary/IHelp64BitSupport.htm
+        # A 32 bit installed.exe can install a 64 bit app, but for it to write to 64bit reg it will
+        # need to use WOW. So the follow is a bit of a guess
```
Comment (twangboy) [inline salt/utils/pkg/win.py]: following

File: salt/utils/pkg/win.py [hunk 14]
```diff
@@ -0,0 +1,1295 @@
...
+
+        if self.__version_only:
+            # package name and package version list, are only info being return
```
Comment (twangboy) [inline salt/utils/pkg/win.py]: are the only info being returned

File: salt/utils/pkg/win.py [hunk 15]
```diff
@@ -0,0 +1,1295 @@
...
+            reg_key_all, _, _, _ = zip(*win32api.RegEnumKeyEx(handle))  # pylint: disable=no-member
+            win32api.RegCloseKey(handle)  # pylint: disable=no-member
+            for reg_key in reg_key_all:
+                self.__collect_software_info(None, reg_key, arch_flag)
+
+        if not user_pkgs:
+            return
+
+        # Process software installed under all USERs, this adds significate processing time.
+        # Their is not 32/64 bit registry redirection under user tree
```
Comment (twangboy) [inline salt/utils/pkg/win.py]: There

File: salt/utils/pkg/win.py [hunk 16]
```diff
@@ -0,0 +1,1295 @@
...
+        log.debug('Processing user software.. please wait')
```
Comment (twangboy) [inline salt/utils/pkg/win.py]: ...

File: salt/utils/pkg/win.py [hunk 17]
```diff
@@ -0,0 +1,1295 @@
...
+    version_only = False
+    if str(sys.argv[1]) == 'list':
+        version_only = True
+    if str(sys.argv[2]) == 'system+user':
+        user_pkgs = True
+    import json
+    import timeit
+
+    def run():
+        """ Main run code when this modules is run directly
```
Comment (twangboy) [inline salt/utils/pkg/win.py]: module

**Review by twangboy — CHANGES_REQUESTED** [review]

**Comment by damon-atkins** [comment]
Thanks for the feedback.
@morganwillcock package id can be anything, however the less unique it is, the more likely for a name clash.  The code supports supply a class and then this class is used to determine the packaged id.

It this stage of its release its more about running on a lot of systems to find the bugs. Even though it may make it into Oxygen, it will not be used by anything. It needs testing on as many computers as possible, being in the Oxygen release will help with that.

@twangboy It's both a library and supports running direct, its not a module.  Before being imported into a salt execution module it needs more people to run it on computers. e.g. different windows languages etc.  Thanks for wording advice/corrections.  At the moment it will be a hidden gem inside Oxygen. So we can get more testers, and then module can be used in a execution module in Fluorine.

**Comment by damon-atkins** [comment]
@morganwillcock  pls try it again on the system which broke the user reporting.   Salt will never be able to manage user installed packages (i.e. you would need runas as part of install/remove). But I include the user information so people could audit PC, and some installers software in System and User space, unfortunately current python installer does this.

@twangboy I have gone through your review. Except I am yet to fix the doc-string `"""` and line feed. Next push.

File: salt/utils/pkg/win.py [hunk 18]
```diff
@@ -0,0 +1,1295 @@
...
+                if exc.winerror == winerror.ERROR_FILE_NOT_FOUND:
+                    # Not Found
+                    log.warning('Not Found {0}\\{1} 32bit {2}'.format(self.__reg_hive,
+                                                                      self.__reg_upgradecode_path,
+                                                                      self.__reg_32bit))
+                    return ''
+                raise
+            squid_upgrade_code_all, _, _, suc_pytime = zip(*win32api.RegEnumKeyEx(uc_handle))  # pylint: disable=no-member
+
+            # Check if we have already scan these upgrade codes before, and also
```
[the remark this answers is not in the document]
Comment (damon-atkins) [reply salt/utils/pkg/win.py]: Next Push

File: salt/utils/pkg/win.py [hunk 19]
```diff
@@ -0,0 +1,1295 @@
...
+        # pylint: enable=no-member
+        if mod_win_time:
+            # at some stage __int__() was removed from pywintypes.datetime to return secs since 1970
+            if hasattr(mod_win_time, 'utctimetuple'):
+                self.__mod_time1970 = time.mktime(mod_win_time.utctimetuple())
+            elif hasattr(mod_win_time, '__int__'):
+                self.__mod_time1970 = int(mod_win_time)
+
+    def __squid_to_guid(self, squid):
+        """Squished GUID (SQUID) to GUID.
```
[the remark this answers is not in the document]
Comment (damon-atkins) [reply salt/utils/pkg/win.py]: Next Push

File: salt/utils/pkg/win.py [hunk 20]
```diff
@@ -0,0 +1,1295 @@
...
+try:
+    from salt.utils.versions import LooseVersion
+except ImportError:
+    from distutils.version import LooseVersion  # pylint: disable=blacklisted-module
+
+
+# pylint: disable=too-many-instance-attributes
+
+class RegSoftwareInfo(object):
+    """
```
[the remark this answers is not in the document]
Comment (damon-atkins) [reply salt/utils/pkg/win.py]: Next Push

File: salt/utils/pkg/win.py (hunk 2 above)
Comment (damon-atkins) [reply salt/utils/pkg/win.py]: Done

**Review by twangboy — CHANGES_REQUESTED** [review]
Waiting for Next Push

**Comment by damon-atkins** [comment]
That should be it, assuming the lint passes.

**Review by morganwillcock — COMMENTED** [review]
I've tested this again with the same Windows image, it successfully got the WiX installer data from the user profile and there was no crash. So I think this is fixed.

**Review by twangboy — APPROVED** [review]

**Pull request merged**